torch/fx/passes/infra/partitioner.py

```diff
@@ -55,7 +56,44 @@ def __is_node_supported(self, node: Node) -> bool:
             self.operator_support.is_node_supported(dict(self.graph_module.named_modules()), node)
         )
 
+    # This function iterates over all the nodes in the graph and starts a new DFS traversal
+    # for each node. During these traversals it maintains two dictionaries one which keeps
+    # track of all the nodes that are encountered in a DFS traversal starting at a particular
+    # node and another dictionary which keeps track of all the nodes that can reach a particular node.
+    def _cache_dfs_path(self):
```

Comment:
Effectively, you are maintaining two "dependency_map" for every pair of nodes in the graph. 

I think this is a very useful tool in general, and it worth refactoring this into a util class. 

```
class DependencyViewer:
    def __init__(self, graph_module):
        self.upstreams = defaultdict(set)  # aka dfs_paths_to_node
        self.downstreams = defaultdict(set) # aka dfs_paths_from_node
        
    def depends_on(node_a, node_b) -> bool: 
        ...
    
    def downstreams_of(node):
         ...

    def upstreams_of(node): 
         ...
```

Reply:
Good idea!

Reply:
I realized I have implemented this before 
https://www.internalfb.com/code/fbsource/[2f71ada0401e]/fbcode/pytorch/pippy/pippy/fx/passes/infra/partitioner.py?lines=48

Somehow this is not master ??!!

Reply:
Haha that's very interesting! Very similar approach.